### PR TITLE
CNN for Digit Recognition Kaggle challenge

### PR DESCRIPTION
**Pull request by PlantsAndBuildings**
Implementation of a CNN that solves the introductory [Digit Recognition challenge](https://www.kaggle.com/c/digit-recognizer) on Kaggle. Using the current setting of hyperparameters, I was  able to get an accuracy of ~~91.185 %~~ 94.485% on the hidden test data.

### CODE REVIEW

**Comment by PlantsAndBuildings** [comment]
I'm sorry, I didn't see that #9 was already open .. should I close this PR?

**Comment by s1998** [comment]
@zoq Since this model is performing better than #9  I think it would be a good idea to merge this instead of that. I could later open a PR for using the same model with MNIST dataset.

**Comment by PlantsAndBuildings** [comment]
@s1998 Really sorry for duplicating your PR.

@zoq I'm fine with merging either as long as @s1998 doesn't have an issue with it (I'm stepping all over his toes with this PR). I would agree with @s1998 that we should merge a model that performs well.

File: Kaggle/DigitRecognizerCNN/src/DigitRecognizerCNN.cpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+  // Generate labels for the test dataset.
+  Row<size_t> testPred = getLabels(testPredOut);
+  cout << "Saving predicted labels to \"results.csv\" ..." << endl;
+
+  // Saving results into Kaggle compatibe CSV file.
+  save("results.csv", "ImageId,Label", testPred);
+  cout << "Results were saved to \"results.csv\". This file can be uploaded to "
+      << "https://www.kaggle.com/c/digit-recognizer/submissions"
+      << endl;
+  cout << "Finished" << endl;
```
Comment (rcurtin) [inline Kaggle/DigitRecognizerCNN/src/DigitRecognizerCNN.cpp]: I know this is really picky but do you want to add some punctuation to the end of this output? :)

File: Kaggle/DigitRecognizerCNN/src/DigitRecognizerCNN.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    Row<size_t> predLabels = getLabels(predOut);
+    double trainAccuracy = accuracy(predLabels, trainY);
+    // Get predictions on validating data points.
+    model.Predict(validX, predOut);
+    // Calculate accuracy on validating data points.
+    predLabels = getLabels(predOut);
+    double validAccuracy = accuracy(predLabels, validY);
+
+    cout << i << " - accuracy: train = "<< trainAccuracy << "%," <<
+        " valid = "<< validAccuracy << "%" <<  endl;
```
Comment (rcurtin) [inline Kaggle/DigitRecognizerCNN/src/DigitRecognizerCNN.cpp]: It might be better to prepend `"Epoch "` before `i`, i.e., `cout << "Epoch " << i << ": "` or something like this.  Let me know what you think.

**Comment by rcurtin** [comment]
Thanks for the contribution!  I do have a big concern with the performance though.  I have been playing with TensorFlow recently and even with a three-layer FFN (not even a CNN) with very simple optimizers I can get 97-98% accuracy.  So I am a little concerned that something is wrong here (maybe not in this code but somewhere else in the library) or perhaps that some tuning can get us much better performance.  It's probably worth investigating in the future.
